### ISSUE TITLE
Can't get an AssetRenditionDispatcher to pick up an alternative config

### ISSUE THREAD

**Issue by toniedzwiedz** [adobe/asset-share-commons#630]
**Describe the bug**
I'm working with Asset Share Commons and I'm aiming to use an alternative rendition mapping to change how thumbnails are handled on a search page.

The URLs rendered by the search component resolve to the `AssetRenditionServlet`, which uses `AssetRenditionDispatcher`s to translate the incoming parameters. There are a few underlying implementations and the one I'm trying to tweak is `InternalRedirectRenditionDispatcherImpl`. 

I've had a look at the implementation and it seems I could achieve this by providing an alternative configuration. This is the one that comes OOTB.
```
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          service.ranking="-9001"
          label="Search Results"
          name="asset-share-commons-search-results"
          hidden="{Boolean}true"
          rendition.mappings="[card=${asset.path}.thumb.319.319.jpg,list=${asset.path}.thumb.140.140.jpg]"
/>
```
What I want to achieve is remap the `card` rendition to something else.

I've created a `sling:OsgiConfig` node under `/apps` that looks roughly like this:

```
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          hidden="true"
          label="My Search Results"
          name="my-search-results"
          rendition.mappings="[card=${asset.path}.preview.400.400.jpg]"
          service.ranking="-8000"/>
```
and I can see the configuration in the Felix console. However, as the Servlet attempts to load the config, the order is not what I would expect it to be.

No matter how high the ranking I set in the node, the instance of `InternalRedirectRenditionDispatcherImpl` backed by this config ends up second (when using negative numbers with lower module ) or last (when using positive numbers) in the list at [this line](https://github.com/adobe/asset-share-commons/blob/asset-share-commons-1.8.0/core/src/main/java/com/adobe/aem/commons/assetshare/content/renditions/impl/AssetRenditionServlet.java#L80). The configuration is read but due to its position in the list, it's ignored by the Servlet (see screenshot below).

Is this a bug or am I misunderstanding the way this should work? Just looking at the underlying ASC and Sling code, specifically the `AssetRenditionsImpl` and `RankedServices` class, it seems a high ranking should allow my config to be picked up as the first one.

**Environment**
* AEM Version: 6.4.8.4
* Asset Share Commons Version: 1.8.0
* Author and Publish

**To Reproduce**
1. Provide an alternative config for `InternalRedirectRenditionDispatcherImpl`
2. Open an image using a URL resolving to the `AssetRenditionServlet`, e.g. http://localhost:4503/content/dam/foo/en/audio/2021/3/animal_cat_meow.mp3.renditions/card/asset.rendition

**Expected behavior**
The suffix allows the servlet to translate the URL to one that matches the rendition mapping of the highest-ranking service.

**Screenshots**
![image](https://user-images.githubusercontent.com/8336085/127319221-dbc223d7-79ab-47e8-9a5e-a2ba22dbe86b.png)

**Additional context**
The reason I'm trying to achieve this is to override the way the default image placeholder is handled for assets that can't have renditions generated automatically. I want to limit this change in behaviour to assets (no affecting pages), only affect assets on pages based on Asset Share Commons (therefore not the Sites console or any other applications installed on the same AEM instance).
ASC ultimately falls back to `com.day.cq.wcm.core.impl.servlets.ThumbnailServlet` which appears to use `com.day.cq.dam.commons.thumbnail.provider.AssetThumbnailProvider`. These seem to remember the days of Day and they aren't configurable. 
I also wouldn't want to add an extra file for every asset there is. I know it's been suggested in https://github.com/adobe/asset-share-commons/issues/526 but it sounds like it would be a pain to update the image if necessary.

I suppose a way I could hack around this would be to have my CRX package overwrite `/apps/asset-share-commons/config/com.adobe.aem.commons.assetshare.content.renditions.impl.dispatchers.InternalRedirectRenditionDispatcherImpl-SearchResults.xml` but that would cause tricky side effects depending on which package gets installed in which order; an ops nightmare.

**Comment by davidjgonzalez** [comment]
Let me check this out - what you're doing looks right, and i swore i helped someone do this same thing a month ago :) ! If ranking isnt supported then def gotta fix it

**Comment by toniedzwiedz** [comment]
Thanks @davidjgonzalez!

**Comment by davidjgonzalez** [comment]
ok! .. that was a weird one :)

So the services are ranked using RankedServices Sling API.
RankedServices requires `service.ranking` to be persisted as an Integer type (it cant be a string that's coerce-able to an Integer, it must be declared AS an Integer.)

So, right now if you go into ConfigMgr and use the WebUI to create a new one, it saves it as a String, and Strings are all evaluated as 0.

So, basically, in the interim, you can make your OSGi config files declare your values as `service.ranking:Integer` with a > 0 value. My PR should otherwise generally fix it - however if you'd want to update any configs youve created/updated to ensure the data type is explicitly an Integer.

**Comment by davidjgonzalez** [comment]
Contains the fix: 

[asset-share-commons.all-2.0.5-SNAPSHOT.zip](https://github.com/adobe/asset-share-commons/files/6902162/asset-share-commons.all-2.0.5-SNAPSHOT.zip)

**Comment by toniedzwiedz** [comment]
Good one indeed :) 

Thanks for looking into this. I'll need to port this fix into ASC 1.8 as that's what my client is using and a major upgrade is off the table for the time being. Do you reckon those configuration changes could be ported 1:1?

What's the process to raise a PR against an older version?

**Comment by davidjgonzalez** [comment]
here pr for 1.x  https://github.com/adobe/asset-share-commons/pull/633 with a snapshot build

**Comment by toniedzwiedz** [comment]
Thanks @davidjgonzalez. I'll install the patch locally and let you know if it works.

However, in order to get this applied properly, I'll probably need an official release. My client is a financial institution and while I could easily build and install a working package, I may have a hard time getting it added as a dependency or otherwise installed in production.

**Comment by toniedzwiedz** [comment]
>So, basically, in the interim, you can make your OSGi config files declare your values as service.ranking:Integer with a > 0 value. My PR should otherwise generally fix it - however if you'd want to update any configs youve created/updated to ensure the data type is explicitly an Integer. 

I've noticed that `Integer` doesn't seem to work when installing the configuration through a CRX package.

Using `Integer` in the JCR property value causes the node creation to fail
```
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          hidden="true"
          label="Public DAM Search Results"
          name="public-dam-search-results"
          rendition.mappings="[card=${asset.path}.preview.319.319.jpg]"
          service.ranking="{Integer}2"/>
```
Output from package installation in CRXDE:
```
E /apps/foo/config/com.adobe.aem.commons.assetshare.content.renditions.impl.dispatchers.InternalRedirectRenditionDispatcherImpl-MySearchResults (org.xml.sax.SAXException: unknown type: Integer java.lang.IllegalArgumentException: unknown type: Integer)
```

and using `Long`

```
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          hidden="true"
          label="Public DAM Search Results"
          name="public-dam-search-results"
          rendition.mappings="[card=${asset.path}.preview.319.319.jpg]"
          service.ranking="{Long}2"/>
```
causes the value to be created as a `String` 🤯 



I've manually edited the config through CRXDE to set `service.ranking` to `2`, using `Long` as the type

![image](https://user-images.githubusercontent.com/8336085/127647042-c27e483b-e110-4353-861b-647ec7d5f344.png)

and restarted the instance but it still appears to load the value as a String without the accompanying ASC code changes from your PR.

![image](https://user-images.githubusercontent.com/8336085/127646697-98e7d2d3-2de5-490a-959e-47cab5574694.png)

**Comment by toniedzwiedz** [comment]
I've also cherry-picked the changes from [the PR](https://github.com/adobe/asset-share-commons/pull/633/files) onto 1.8.0, built the bundle and deployed it. I can see the values are represented as Integers but the order still looks wrong.

![image](https://user-images.githubusercontent.com/8336085/127652436-ee8628c4-7c78-487f-9584-67c1aff88260.png)

which is quite confusing.

Meanwhile, an alternative workaround has been suggested to me. The OOTB config is saved under `/apps/asset-share-commons/config`. I could have two copies (somewhat redundant) with the relevant value overridden maintained for a more specific runmode, e.g. `/apps/foo/config/author` and `/apps/foo/config/publish`

This appears to work as expected and the only downside is minor duplication.

**Comment by davidjgonzalez** [comment]
sunuva gun .. https://issues.apache.org/jira/browse/SLING-2477

Looks like we have to change them from sling:osgiConfig to .config or something for 1.x line .. good call.

**Comment by toniedzwiedz** [comment]
Anyway, it looks like I have a functioning workaround (author and publish-specific config that trumps the OOTB one unassociated with runmodes) for the time being so I can proceed with the work this was blocking.

> Also, if its NOT an integer, the value defaults to 0 ... so Longs wont work

That explains the behaviour I've been seeing for the last hour or so where the order was... random, despite my efforts. All zeros apparently, despite the values I could see with the debugger. Makes me want to set this to 

```
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          label="Search Results"
          name="asset-share-commons-search-results-publish-specific"
          hidden="{Boolean}true"
          rendition.mappings="[card=${asset.path}.preview.319.319.jpg]"
          service.ranking="This isn't an actual ranking due to a bug, all numbers placed here will default to 0. Use runmodes."/>
```
:D 

What a day.

**Comment by davidjgonzalez** [comment]
Ugh - damn - sorry :( ... i think we if we switch to config we can specify Integers. I dont think? that .cfg.json is supported for everything 1.x supports (i wish there was a support matrix for all this stuff, and their permutations) ..

**Comment by toniedzwiedz** [comment]
> Ugh - damn - sorry :( 
No worries, having two config files instead of one is a very low price compared to some of the things I was considering above. Also, this has been super interesting ;)

As far as ASC itself is concerned, I still have a very limited understanding of how each individual feature works. `.cfg.json` sound more in line with AEMaaCS, which I understand is supported by 2.x.

For 1.x, `.config` files should probably be fine.

**Comment by davidjgonzalez** [comment]
@toniedzwiedz ok - .config should be used in: https://github.com/adobe/asset-share-commons/pull/633

I put a snapshot of ui.apps if you have a second to check it out. I debugged  and looks good from there as well:

![2021-07-30 at 5 02 PM](https://user-images.githubusercontent.com/1451868/127711247-f6975170-0c6f-4a34-be94-74b82ca0c6fa.jpg)

**Comment by toniedzwiedz** [comment]
@davidjgonzalez sure, I'll try and have a look.
